README.md
# @wufengteam/core

## 组件组册

```ts
import { Button } from 'xxx';

const platform = 'pc'; // or h5 手动指定注册的平台，留空使用默认。

wufengController.registerComponent(
  Button,
  {
    todoProps: {},
    label: '按钮',
    type: 'Button',
    description: '描述',
    image: '展示大图',
    groupsName: '容器',
    todoStyles: {},
    todoEvents: [],
    todoActionList: [],
    props: {},
  },
  platform,
);
```

## 组件使用

### 获取配置

```ts
const platform = 'pc'; // or h5 手动指定注册的平台，留空使用默认。
const config = wufengController.getComponentConfigByType('Button', platform);

// 这里单纯的取到配置数据，用于项目中的数据流传递
pageData.push({ ...config, id: 12312 });
```

### 获取组件

```tsx
const config = findItem(pageData, (i) => i.id === id);
const Button = wufengController.getComponentByType(
  config.type,
  config.platform,
);
// 这里单纯的取到渲染的组件 Dom，用于项目中的组件渲染
return <Button />;
```

## 设置当前注册器平台

组件增加 platform 属性，优先级最高，次一次取调用 api 传入的 platform，最后兜底取默认的 wufengController.platform

```ts
wufengController.setPlatform('h5'); // 切换 pc 和 h5
```


CHANGELOG.md
# Change Log

## 2.7.14

增加订阅服务

## 2.7.13

### Patch Changes

- chore: vp

## 2.7.12

### Patch Changes

- 9177d994: 低零融合接口路径调整
- c7c6d5ae: 目录切换新接口

## 2.7.11

### Patch Changes

- fix: register again warn

## 2.7.10

### Patch Changes

- feat: add setting position

## 2.7.9

### Patch Changes

- chore: vp

## 2.7.8

### Patch Changes

- chore: vp

## 2.7.7

### Patch Changes

- chore: vp

## 2.7.6

### Patch Changes

- 9ef873b3: feat: 增加 compType 组件类型为仪表组件的定义

## 2.7.5

### Patch Changes

- feat: input add hasFormItem

## 2.7.4

### Patch Changes

- chore: vp

## 2.7.3

### Patch Changes

- chore: vp

## 2.7.2

### Patch Changes

- chore: vp

## 2.7.1

### Patch Changes

- chore: vp

## 2.7.0

### Minor Changes

- chore: vp

## 2.6.5

### Patch Changes

- chore: vp

## 2.6.2

### Patch Changes

- f4b86ce2: chore: vp

## 2.6.1

### Patch Changes

- chore: vp

## 2.6.0

### Minor Changes

- feat: support ie

## 2.5.9

### Patch Changes

- feat: support ie

## 2.5.8

### Patch Changes

- chore: vp

## 2.5.7

### Patch Changes

- 2ec03b4: feat: support browser
- 623fb6d: 添加函数计算

## 2.5.6

### Patch Changes

- chore: vp

## 2.5.5

### Patch Changes

- chore: vp

## 2.5.4

### Patch Changes

- chore: vp

## 2.5.3

### Patch Changes

- chore: vp

## 2.5.2

### Patch Changes

- chore: vp
- chore: vp

## 2.5.1

### Patch Changes

- chore: vp

## 2.5.0

### Minor Changes

- chore: vp

## 2.4.8

### Patch Changes

- feat: add children

## 2.4.7

### Patch Changes

- chore: vp

## 2.4.6

### Patch Changes

- chore: code types

## 2.4.5

### Patch Changes

- feat: add targetType

## 2.4.4

### Patch Changes

- 9b5fbbe: feat: version

## 2.4.3

### Patch Changes

- fix: moment

## 2.4.2

### Patch Changes

- 6a78e44: chore: compType add 7

## 2.4.1

### Patch Changes

- 2ef3f45: feat: cloud components

## 2.4.0

### Minor Changes

- 8219b5a: fix: core working in browser

## 2.3.0

### Minor Changes

- 2d39381: feat: add register actions

### Patch Changes

- 8bda813: feat: add loadable component

## 2.2.0

### Minor Changes

- 478600f: feat: register icons

### Patch Changes

- 4364c77: fix: core setEvents schema
- chore: add isInlineBlock

## 2.1.2

### Patch Changes

- feat: core 增加 platform

## 2.1.1

### Patch Changes

- 7760c32: feat: 组件规则修改

All notable changes to this project will be documented in this file. See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 2.1.0 (2022-02-22)

### Features

- changeset ([d0c5c94](https://git-nj.iwhalecloud.com/zerocodeplatformtest/nocode/commits/d0c5c944cbf4deec1e3ccb7e2833f862e3ee6cca))
- core schema ([513db23](https://git-nj.iwhalecloud.com/zerocodeplatformtest/nocode/commits/513db23b92623631964990187b67369140edfaf9))
- init wufeng core ([acd14b0](https://git-nj.iwhalecloud.com/zerocodeplatformtest/nocode/commits/acd14b029810ae163209770440bcd3a9a180a632))
- version packages ([6c5a951](https://git-nj.iwhalecloud.com/zerocodeplatformtest/nocode/commits/6c5a9518e47f199faf5ec75595589a3096439205))
- wufeng core ([515296d](https://git-nj.iwhalecloud.com/zerocodeplatformtest/nocode/commits/515296df05d6e9b8c7d22d727b2bc3a43d4532ba))

# @wufengteam/core

## 2.0.0

### Major Changes

- 8c058f1: feat: 增加统一注册中心的校验规则，不仅仅是 types 层面的校验，加入 joi 校验，在 js 环境中也保持校验有效。

  1、注册时自动校验 2、手动使用 schemaComponent 和 schemaInput

  ```
  const { error } = wufengController.schemaComponent.validate(options);
  ```

- acd14b0: feat: 根据低代码的规则修改中心注册器的定义

## 1.1.1

### Patch Changes

- init changeset


src/index.test.ts
import { wufengController } from './index';

test('inputComponents', () => {
  wufengController.registerInput(
    {},
    {
      type: 'select',
      istodoBind: true,
      label: 'jiaobu',
      groupsName: '基础',
      rules: [],
      props: {
        min: '',
        max: '',
        label: '名称',
        options: [],
        dependProps: {},
        inhibitBtn: 'true',
      },
      defaultValue: '',
    },
  );
  const current = wufengController.getInputConfigByType('select');
  expect(current?.type).toBe('select');
});

test('inputComponents error', () => {
  const a = wufengController.registerInput({}, {
    type: 'abc',
    istodoBind: true,
    label: 'jiaobu',
    abc: '基础',
    rules: [],
    props: {
      min: '',
      max: '',
      label: '名称',
      options: [],
      dependProps: {},
      inhibitBtn: 'true',
    },
    defaultValue: '',
  } as any);
  const current = wufengController.getInputConfigByType('abc');
  expect(current).toBe(null);
  expect(a?.message).toBe('"abc" is not allowed');
});

test('Components', () => {
  wufengController.registerComponent(
    {},
    {
      todoProps: {
        footer: {
          type: 'select',
          istodoBind: true,
          label: 'jiaobu',
          groupsName: '基础',
          rules: [],
          props: {
            min: '',
            max: '',
            label: '名称',
            options: [],
            dependProps: {},
            inhibitBtn: 'true',
          },
        },
      },
      label: '按钮',
      type: 'Button',
      description: '描述',
      image: '展示大图',
      groupsName: '容器',
      todoStyles: {
        themeColor: {
          type: 'ColorPicker',
          label: '颜色选择器',
          defaultValue: {},
          props: { options: [] },
        },
      },
      todoEvents: [
        {
          label: '值改变时回调',
          value: 'onChange',
          params: [
            { title: '变化值', value: '$e.target.value$', name: 'vals' },
          ],
        },
      ],
      todoActionList: ['setMobileValue'],
      props: {
        header: 'a123',
      },
    },
  );
  const current = wufengController.getComponentConfigByType('Button');
  expect(current?.description).toBe('描述');
});

test('label setLang', () => {
  const defaultLabels = {
    type: '类型',
  };
  wufengController.pushLabels(defaultLabels, 'zh-CN');
  const text = wufengController.findLabel('type');
  expect(text).toBe('类型');
  wufengController.pushLabels(
    {
      type: 'Type text',
    },
    'en-UK',
  );
  wufengController.setLang('en-UK');
  const current = wufengController.findLabel('type');
  expect(current).toBe('Type text');
});

test('schemaInput success', async () => {
  const { error } = wufengController.schemaInput.validate({
    type: 'select',
    istodoBind: true,
    label: 'jiaobu',
    groupsName: '基础',
    rules: [],
    props: {
      min: '',
      max: '',
      label: '名称',
      options: [],
      dependProps: {},
      inhibitBtn: 'true',
    },
    defaultValue: '',
  });
  expect(error?.message).toBe(undefined);
});


src/emitter.test.ts
import { wufengController } from './index';

test('emitter', () => {
  const fn = jest.fn();
  wufengController.useSubscription('core-test', (val: any) => {
    fn(val);
  });
  wufengController.emit('core-test', 'Hello!');
  expect(fn).toHaveBeenCalled();
  expect(fn).toHaveBeenCalledWith('Hello!');
});


src/schema/index.test.ts
import { schemaInput, schemaComponent } from './index';

test('schemaInput error', async () => {
  // @ts-ignore
  const { error } = schemaInput.validate({ type: 1994 });
  expect(error?.message).toBe('"type" must be a string');
});

test('schemaInput success', async () => {
  const { error } = schemaInput.validate({
    type: 'select',
    istodoBind: true,
    label: 'jiaobu',
    groupsName: '基础',
    rules: [],
    props: {
      min: '',
      max: '',
      label: '名称',
      options: [],
      dependProps: {},
      inhibitBtn: 'true',
    },
    defaultValue: '',
  });
  expect(error?.message).toBe(undefined);
});

test('schemaComponent success', async () => {
  const { error } = schemaComponent.validate({
    todoProps: {
      footer: {
        type: 'select',
        istodoBind: true,
        label: 'jiaobu',
        groupsName: '基础',
        rules: [],
        props: {
          min: '',
          max: '',
          label: '名称',
          options: [],
          dependProps: {},
          inhibitBtn: 'true',
        },
        defaultValue: '',
      },
    },
    label: '按钮',
    type: 'Button',
    compType: 99,
    description: '',
    image: '展示大图',
    groupsName: '容器',
    todoStyles: {
      themeColor: {
        type: 'ColorPicker',
        label: '颜色选择器',
        defaultValue: {},
        props: { options: [] },
      },
    },
    todoEvents: [
      {
        label: '值改变时回调',
        value: 'onChange',
        params: [{ title: '变化值', value: '$e.target.value$', name: 'vals' }],
      },
    ],
    todoActionList: ['setMobileValue'],
    props: {
      header: 'a123',
    },
  });
  expect(error?.message).toBe(undefined);
});


src/types.ts
export type JSONValue = any;

export type JSONObject = Record<string, JSONValue>;

export type JSONArray = JSONValue[];

export type InputObject = Record<string, Input>;

export interface Input {
  /**
   * 类型，唯一标识，类型重复无法注册
   */
  type: string;
  /**
   * 是否允许绑定数据源
   */
  istodoBind?: boolean;
  /**
   * 显示给用户的文本
   */
  label?: string;
  /**
   * 分组名称，直接按字符匹配分组，便于直接扩展
   */
  groupsName?: string;
  /**
   * 值校验规则，正则表达式
   */
  rules?: string[];
  /**
   * 直接透传给属性编辑器的属性，用于属性编辑器的复用情况
   */
  props?: JSONObject;
  /**
   * 属性编辑器的默认值
   */
  defaultValue?: JSONValue;
  /**
   * 标记这个设置器最终修改的属性对象，比如可能展示的时候，是在 todoStyle 中，但是修改的是属性的 props 值
   */
  targetType?: string;
  /**
   * 设置器在面板中展示的位置
   */
  /**
   * 属性变化前的hook，可以操作当前节点的字段。
   */
  onBeforeChange?: (options?: any, extraProps?: any) => Promise<any>;
  /**
   * 组件是否自带 Form.Item
   */
  hasFormItem?: boolean;
  /**
   * 设置器在面板中展示的位置
   */
  position?: string;
}

export interface Component {
  /**
   * 组件名称
   */
  label?: string;
  /**
   * 类型，唯一标识，类型重复无法注册
   */
  type: string;
  /**
   * 组件关联平台，这里定义的优先级最高
   */
  platform?: string;
  /**
   *  组件类型  0：容器  1：通用  2：数据录入  3：数据展示 4：反馈  5：导航  6：布局 7:图标  99:订阅组件 88: 可视化仪表盘 66: 低零融合组件
   */
  compType?: number;
  /**
   * 基础组件使用较多，组件所在库标识
   */
  compLib?: string;
  /**
   * 组件允许编辑的所有属性
   */
  todoProps: InputObject;
  /**
   * 组件的所有属性默认值，包括允许编排的属性和其他属性
   */
  props?: JSONValue;
  /**
   * 允许编辑的所有样式
   */
  todoStyles?: InputObject;
  /**
   * 组件的所有样式默认值，包括允许编排的样式和其他样式
   */
  style?: JSONValue;
  /**
   * 是否为容器组件
   */
  isContainer?: boolean;
  /**
   * 标识为 InlineBlock 布局，编辑器需要特殊处理
   * 比如 拖入一个文本，会默认占用一行，需要修改外层容器的样式
   */
  isInlineBlock?: boolean;
  /**
   * 是否可以放置业务对象
   */
  isBusiObjContainer?: boolean;
  /**
   * 是App表单子项
   */
  isAppChildForm?: boolean;
  /**
   * 组件详情，用于 card 显示，或资产商店说明
   */
  description?: string;
  /**
   * 组件大图
   */
  image?: string;
  /**
   * 左侧编辑时，配套显示的 icon
   * 允许直接写 http:// 和 icon type
   */
  icon?: string;
  /**
   * 分组名称，直接按字符匹配分组，便于直接扩展
   */
  groupsName?: string;
  /**
   * 允许编辑的事件
   */
  todoEvents?: JSONArray;
  /**
   * 初始化时默认绑定的事件
   */
  setEvents?: JSONArray;
  /**
   * 组件暴露的所有事件
   */
  todoActionList?: string[];
  /**
   * 需要对标的 api，比如编辑器中配置 button 的 value，真实的 button 值属性是 children
   * transform: {
   * // 需要翻译的字段名称
   * value: 'children',
   * },
   */
  transform?: JSONArray;
  /**
   * 真实的组件文件，再调用注册时会被绑定
   * 删除实例中的 config，通过 getComponentByType 获取
   */
  // class?: string;
  /**
   * 子组件，用于组合组件
   */
  components?: Component[];
  /**
   * 只有一个子组件
   */
  children?: any;
  /**
   * app校验字段
   */
  validator?: any;
  /**
   * 允许用户自定义事件，可能没用，待确认
   */
  customEvent?: boolean;
  /**
   * 标记组件能够存放的父级元素
   * DFormCustom 的父级容器只能是 "DForm", "DGroup"
   * {
   *   type: "DFormCustom",
   *   onlyRoot: ["DForm", "DGroup"]
   * }
   */
  onlyRoot?: string[];
}

export interface FunctionDefine {
  // 函数唯一标识
  type: string;
  // 函数名称
  name: string;
  // 函数描述
  description?: string;
  // 函数文档
  docs?: string;
  // 分组名字
  groupsName?: string | string[];
}


src/schema/index.ts
import Joi from '@alita/joi';
import type {
  ObjectSchema,
  ValidationResult,
  ValidationOptions,
  AsyncValidationOptions,
} from '@alita/joi';
import type { Input, Component, FunctionDefine } from '../types';

// 扩展 Joi 的类型定义，校验值的时候，可以有智能提醒，可以在编写代码的时候就提示错误
interface ObjectCustomSchema<TSchema = any> extends ObjectSchema<TSchema> {
  // eslint-disable-next-line
  validate(
    value: TSchema,
    options?: ValidationOptions,
  ): ValidationResult<TSchema>;
  // eslint-disable-next-line
  validateAsync(value: TSchema, options?: AsyncValidationOptions): Promise<any>;
}

const schemaInput = Joi.object({
  /**
   * 类型，唯一标识，类型重复无法注册
   */
  type: Joi.string(),
  /**
   * 是否允许绑定数据源
   */
  istodoBind: Joi.boolean(),
  /**
   * 显示给用户的文本
   */
  label: Joi.string(),
  /**
   * 分组名称，直接按字符匹配分组，便于直接扩展
   */
  groupsName: Joi.string(),
  /**
   * 值校验规则，正则表达式
   */
  rules: Joi.array().items(Joi.string()),
  /**
   * 直接透传给属性编辑器的属性，用于属性编辑器的复用情况
   */
  props: Joi.object(),
  /**
   * 属性编辑器的默认值
   */
  defaultValue: Joi.any(),
  /**
   * 标记这个设置器最终修改的属性对象，比如可能展示的时候，是在 todoStyle 中，但是修改的是属性的 props 值
   */
  targetType: Joi.string().valid('props', 'styles'),
  /**
   * 属性变化前的hook，可以操作当前节点的字段。
   */
  onBeforeChange: Joi.func(),
  /**
   * 组件是否自带 Form.Item
   */
  hasFormItem: Joi.boolean(),
  /*
   * 设置器在面板中展示的位置
   */
  position: Joi.string().valid('normal', 'left', 'right'),
}) as ObjectSchema<Input>;

const schemaComponent = Joi.object({
  /**
   * 类型，唯一标识，类型重复无法注册
   */
  type: Joi.string(),
  /**
   * 组件名称
   */
  label: Joi.string(),
  /**
   * 组件关联平台，这里定义的优先级最高
   */
  platform: Joi.string(),
  /**
   *  组件类型  0：容器  1：通用  2：数据录入  3：数据展示 4：反馈  5：导航  6：布局 7:图标  99:订阅组件 88: 可视化仪表盘 66: 低零融合组件
   */
  compType: Joi.number().valid(0, 1, 2, 3, 4, 5, 6, 7, 88, 99, 66),
  /**
   * 基础组件使用较多，组件所在库标识
   */
  compLib: Joi.string(),
  /**
   * 组件允许编辑的所有属性
   */
  todoProps: Joi.object().pattern(Joi.string(), schemaInput),
  /**
   * 组件的所有属性默认值，包括允许编排的属性和其他属性
   */
  props: Joi.object(),
  /**
   * 允许编辑的所有样式
   */
  todoStyles: Joi.object().pattern(Joi.string(), schemaInput),
  /**
   * 组件的所有样式默认值，包括允许编排的样式和其他样式
   */
  style: Joi.object(),
  /**
   * 是否为容器组件
   */
  isContainer: Joi.boolean(),
  /**
   * 标识为 InlineBlock 布局，编辑器需要特殊处理
   * 比如 拖入一个文本，会默认占用一行，需要修改外层容器的样式
   */
  isInlineBlock: Joi.boolean(),
  /**
   * 是否可以放置业务对象
   */
  isBusiObjContainer: Joi.boolean(),
  /**
   * 是App表单子项
   */
  isAppChildForm: Joi.boolean(),
  /**
   * 组件详情，用于 card 显示，或资产商店说明
   */
  description: Joi.string().empty(''),
  /**
   * 组件大图
   */
  image: Joi.string().empty(''),
  /**
   * 左侧编辑时，配套显示的 icon
   * 允许直接写 http:// 和 icon type
   */
  icon: Joi.string().empty(''),
  /**
   * 分组名称，直接按字符匹配分组，便于直接扩展
   */
  groupsName: Joi.string(),
  /**
   * 允许编辑的事件
   */
  todoEvents: Joi.array().items(Joi.any()),
  /**
   * 初始化时默认绑定的事件
   */
  setEvents: Joi.array().items(Joi.any()),
  /**
   * 组件暴露的所有事件
   */
  todoActionList: Joi.array().items(Joi.string()),
  /**
   * 需要对标的 api，比如编辑器中配置 button 的 value，真实的 button 值属性是 children
   * transform: {
   * // 需要翻译的字段名称
   * value: 'children',
   * },
   */
  transform: Joi.object(),
  /**
   * 子组件，用于组合组件
   */
  components: Joi.any(),
  /**
   * 大概率和 components 一样只是取个别名
   */
  children: Joi.any(),
  /**
   * app校验字段
   */
  validator: Joi.any(),
  /**
   * 允许用户自定义事件，可能没用，待确认
   */
  customEvent: Joi.boolean(),
  /**
   * 标记组件能够存放的父级元素
   * DFormCustom 的父级容器只能是 "DForm", "DGroup"
   * {
   *   type: "DFormCustom",
   *   onlyRoot: ["DForm", "DGroup"]
   * }
   */
  onlyRoot: Joi.array().items(Joi.string()),
}) as ObjectCustomSchema<Component>;

const schemaFunction = Joi.object({
  /**
   * 函数名称，唯一标识
   */
  type: Joi.string(),
  /**
   * 暂时跟 type 一样
   */
  name: Joi.string(),
  /**
   * 函数简单描述
   */
  description: Joi.string(),
  /**
   * 函数文档
   */
  docs: Joi.string().empty(''),
  /**
   * 函数分组名字，如果属于多个分组可以用数组
   */
  groupsName: [Joi.string(), Joi.array().items(Joi.string())],
}) as ObjectCustomSchema<FunctionDefine>;

export { schemaComponent, schemaInput, schemaFunction, ObjectCustomSchema };
